pmlc/conversion/gpu_to_comp/gpu_to_comp.cc

```diff
@@ -105,121 +113,171 @@ RewriteLaunchFunc::matchAndRewrite(gpu::LaunchFuncOp op,
         "Expected containing function to supply an execution device");
     return mlir::failure();
   }
-  mlir::Value device = func.getArgument(0);
   // Create execution environment.
+  mlir::Value device = func.getArgument(0);
   auto execEnvOp =
       rewriter.create<comp::CreateExecEnv>(loc, execEnvType, device);
   mlir::Value execEnv = execEnvOp.getResult();
+  // Collect consecutive gpu.launch_func Ops starting from current op.
+  std::vector<gpu::LaunchFuncOp> launchOps;
+  if (mlir::failed(getConsecutiveOps<gpu::LaunchFuncOp>(op, launchOps)))
+    return mlir::failure();
+
+  std::map<mlir::Value, mlir::Value, valueComparator> bufferPool;
+  std::vector<std::vector<mlir::Value>> newOperands(launchOps.size());
   // Allocate memory on device for memory operands.
-  std::vector<mlir::Value> newOperands;
-  if (mlir::failed(allocateDeviceMemory(rewriter, loc, execEnv, op.operands(),
-                                        newOperands)))
+  if (mlir::failed(allocateDeviceMemory(rewriter, loc, execEnv, launchOps,
+                                        newOperands, bufferPool)))
     return mlir::failure();
-  mlir::Value funcEvent;
-  if (mlir::failed(createScheduleFuncOp(rewriter, loc, execEnv, op, newOperands,
-                                        funcEvent)))
+  std::vector<mlir::Value> funcEvents;
+  // Create schedule functions within the same execution environment.
+  if (mlir::failed(createScheduleFuncOps(rewriter, loc, execEnv, launchOps,
+                                         newOperands, funcEvents)))
     return mlir::failure();
   // Read device memory back to host.
-  if (mlir::failed(readDeviceMemory(rewriter, loc, execEnv, op.operands(),
-                                    newOperands, funcEvent)))
+  if (mlir::failed(readDeviceMemory(rewriter, loc, execEnv, funcEvents,
+                                    newOperands, bufferPool)))
     return mlir::failure();
   // Deallocate device memory.
-  if (mlir::failed(deallocateDeviceMemory(rewriter, loc, execEnv, op.operands(),
-                                          newOperands)))
+  if (mlir::failed(deallocateDeviceMemory(rewriter, loc, execEnv, bufferPool)))
     return mlir::failure();
   // Destroy execution environment.
   rewriter.create<comp::DestroyExecEnv>(loc, execEnv);
-  // Remove original launch operation.
-  rewriter.eraseOp(op.getOperation());
+  // Remove original launch operations.
+  for (auto launchOp : launchOps) {
+    rewriter.eraseOp(launchOp.getOperation());
+  }
+  return mlir::success();
+}
 
+template <typename T>
+mlir::LogicalResult
+RewriteLaunchFunc::getConsecutiveOps(T op, std::vector<T> &ops) const {
+  // Collect consecutive ops of type T starting from a given op.
+  auto operation = op.getOperation();
+  while (operation != nullptr) {
```

Comment:
This too could be simplified, right?
```
while op isa T:
  ops.push_back(op)
  op = op->getNextNode()
return mlir::success()
```

Reply:
Yes, I will simplify this. Maybe it is better I make the update in the vulkan PR.

Reply:
Changed in [PR1605](https://github.com/plaidml/plaidml/pull/1605/files#diff-6663ff64acb7df7101c5a996e292cb741f317686d2130b8a3a11a24946d8b724R158)